internal/app/synchronizer/synchronizer_service.go

```diff
@@ -46,6 +132,165 @@ func (s *synchronizerService) Register(ctx context.Context, req *pb.RegisterRequ
+	// be in progress). The evaluation routine tracks the completion of all of
+	// these and then changes the state of the synchronizer to idle.
+	results := s.fetchResults(req.Id)
+	return &ipb.EvaluateProposalsResponse{Match: results}, nil
+}
+
+func (s *synchronizerService) addProposals(id string, proposals []*pb.Match) error {
+	s.stateMutex.Lock()
+	defer s.stateMutex.Unlock()
+
+	if !s.proposalCollection() {
+		return status.Error(codes.DeadlineExceeded, "synchronizer currently not accepting match proposals")
+	}
+
+	if _, ok := s.cycleData.requests[id]; !ok {
+		return status.Error(codes.DeadlineExceeded, "request not present in the current synchronization cycle")
+	}
+
+	if len(s.cycleData.requests[id].proposals) > 0 {
+		// We do not currently support submitting multiple separate proposals for a registered request.
+		return status.Error(codes.Internal, "multiple proposal addition for a request not supported")
+	}
+
+	// Proposal can be added for this request so increment the pending requests
+	s.cycleData.pendingRequests.Add(1)
+
+	// Copy the proposals being added to the request data for the specified id.
+	s.cycleData.requests[id].proposals = make([]*pb.Match, len(proposals))
+	copy(s.cycleData.requests[id].proposals, proposals)
+	return nil
+}
+
+func (s *synchronizerService) proposalCollection() bool {
+	return s.state == proposalCollection || s.state == requestRegistration
+}
+
+func (s *synchronizerService) fetchResults(id string) []*pb.Match {
+	t := time.Now()
+	var results []*pb.Match
+
+	// fetchMatches will block till evaluation is in progress. The resultsReady channel
+	// is closed only when evaluation is complete. This is used to signal all waiting
+	// requests that the results are ready.
+	<-s.cycleData.resultsReady
+	results = make([]*pb.Match, len(s.cycleData.requests[id].results))
+	copy(results, s.cycleData.requests[id].results)
+	synchronizerServiceLogger.Debugf("Results ready for id: %v, results: %v, elapsed time: %v", id, getMatchIds(results), time.Since(t).String())
+	return results
+}
+
+// Evaluate aggregates all the proposals across requests and calls the user configured
+// evaluator with these. Once evaluator returns, it copies results back in respective
+// request data objects and then signals the completion of evaluation. It then waits
+// for all the result processing (by fetchMatches) to be compelted before changing the
+// synchronizer state to idle.
+// NOTE: This method is always called while holding the synchronizer state mutex.
+func (s *synchronizerService) Evaluate() {
+	var aggregateProposals []*pb.Match
+	proposalMap := make(map[string]string)
+	for id, data := range s.cycleData.requests {
+		aggregateProposals = append(aggregateProposals, data.proposals...)
+		for _, m := range data.proposals {
+			proposalMap[m.MatchId] = id
+		}
+	}
+
+	synchronizerServiceLogger.Infof("Requesting evaluation of %v proposals", len(aggregateProposals))
```

Comment:
TODO for prom metric here?

Reply:
Created an issue listing some interesting metrics to be added for synchronizer - including this one. Thanks!